v2/yarpcrouter/router.go

```diff
@@ -72,6 +74,56 @@ func NewMapRouter(defaultService string, rs []yarpc.TransportProcedure) MapRoute
 	return router
 }
 
+// EncodingToTransportProcedures converts encoding-level procedures to transport-level procedures.
+func EncodingToTransportProcedures(encodingProcedures []yarpc.EncodingProcedure) ([]yarpc.TransportProcedure, error) {
+	transportProcedures := make([]yarpc.TransportProcedure, len(encodingProcedures))
+	for i, p := range encodingProcedures {
+		var transportHandlerSpec yarpc.TransportHandlerSpec
+		switch p.HandlerSpec.Type() {
+		case yarpc.Unary:
+			transportHandlerSpec = yarpc.NewUnaryTransportHandlerSpec(&unaryTransportHandler{p})
+		default:
+			return nil, fmt.Errorf("unknown handler spec type: %v", p.HandlerSpec.Type())
+		}
+
+		transportProcedures[i] = yarpc.TransportProcedure{
+			Name:        p.Name,
+			Service:     p.Service,
+			HandlerSpec: transportHandlerSpec,
+			Encoding:    p.Encoding,
+			Signature:   p.Signature,
+		}
+	}
+
+	return transportProcedures, nil
+}
+
+// Allows encoding-level procedures to act as transport-level procedures.
+type unaryTransportHandler struct {
+	h yarpc.EncodingProcedure
+}
+
+func (u *unaryTransportHandler) Handle(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
+	res := &yarpc.Response{}
+	decodedBody, err := u.h.Codec.Decode(reqBuf)
+	if err != nil {
+		return res, nil, err
+	}
+
+	body, err := u.h.HandlerSpec.Unary().Handle(ctx, decodedBody)
+	if err != nil {
+		res.ApplicationError = true
+		return res, nil, err
```

Comment:
A little strange to return a mix of non-`nil` and `nil` values for responses, along with a non-`nil` `error`. I understand we might need the `Response` information though. Could you shed more light here? 

In other words, do we always need to return a non-`nil` `Response`? If so, it seems fairly prone to error since we normally discount all return values in the presence of a non-`nil` `error`.

Reply:
I agree, it's strange.

It's also what yarpcjason does currently, it seems: https://sourcegraph.uberinternal.com/github.com/yarpc/yarpc-go@spring/-/blob/v2/yarpcjson/inbound.go#L75-81

I'm not 100% sure how to proceed.